src/pudl/helpers.py

```diff
@@ -916,24 +919,36 @@ def convert_cols_dtypes(df, data_source, name=None):
         :mod:`pudl.constants` ``column_dtypes`` dictionary.
 
     """
+    source_dtypes = pc.column_dtypes[data_source]
     # get me all of the columns for the table in the constants dtype dict
     col_dtypes = {col: col_dtype for col, col_dtype
-                  in pc.column_dtypes[data_source].items()
+                  in source_dtypes.items()
                   if col in list(df.columns)}
 
     # grab only the boolean columns (we only need their names)
     bool_cols = {col: col_dtype for col, col_dtype
                  in col_dtypes.items()
                  if col_dtype == pd.BooleanDtype()}
+    # grab all of the non boolean columns
+    non_bool_cols = {col: col_dtype for col, col_dtype
+                     in col_dtypes.items()
+                     if col_dtype != pd.BooleanDtype()}
     # Grab only the string columns...
     string_cols = {col: col_dtype for col, col_dtype
                    in col_dtypes.items()
                    if col_dtype == pd.StringDtype()}
 
-    # grab all of the non boolean columns
-    non_bool_cols = {col: col_dtype for col, col_dtype
-                     in col_dtypes.items()
-                     if col_dtype != pd.BooleanDtype()}
+    # check for non-nullable string dtypes. we need them to all be the nullable
+    # type, otherwise string cols w/ nulls will not be able to be assigned
+    non_nullable_string_cols = [
+        c for c in source_dtypes.keys()
+        if (source_dtypes[c] == 'string')
+        & (source_dtypes[c] != pd.StringDtype())
+    ]
```

Comment:
Not sure I get what's going on here.  If the dtype is `string` then it is a `pd.StringDtype` "Non-nullable strings" should come back as dtype `object` So I think this must axiomatically be an empty list, no?

Reply:
this is not the list of dtypes from the df. it is the dtypes we have defined in constants

Reply:
Okay I'm definitely confused. I don't understand what this check is attempting to do. Isn't this equivalent to saying something like:
```py
if (x == 1) & (x != 2)
```
* If x == 1 then it must be true that x != 2. (TRUE)
* If x == 2 then both statements are False (FALSE)
* If x == 0 (i.e. it's not 1 and it's not 2) then the first statement is False and the 2nd statement is True (FALSE)

Which is equivalent to just saying `if (x == 1)` isn't it?

Or are you counting on the value of `pd.StringDtype()` be cast to a string when it is compared with a string? Usually for type checking in Python you'd use `isinstance()`

Reply:
Ah, I see why this is confusing. When I select for "strings" it returns all of the string types, including the nullable type. So I made an assumption that I could be more specific about the string type but I WAS WRONG.

So idk how to test this.

```
source_dtypes = {
    'col1': pd.StringDtype(),
    'col2': 'string'
}
# check for non-nullable string dtypes. we need them to all be the nullable
# type, otherwise string cols w/ nulls will not be able to be assigned
string_cols = [
    c for c in source_dtypes.keys()
    if (source_dtypes[c] == 'string')
]
string_cols
```
returns:
```
['col1', 'col2']
```

aaaand
```
nullable_string_cols = [
    c for c in source_dtypes.keys()
    if (source_dtypes[c] == pd.StringDtype())
]
non_nullable_string_cols
```
returns:
```
['col1', 'col2']

Reply:
The string representation of `pd.StringDtype()` is `string`. These two things are referring to the same case. So `pd.StringDtype() == "string"` is axiomatically true, since in the context of the `==` operator the `pd.StringDtype()` class will present itself as its string representation.

I don't understand what you are attempting to check for here.